### PR TITLE
Migrate LlamaDecoderLayer to NNX

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

This is co-work with @bvandermoon and @shuningjin 

Example to migrate llama2 to NNX. This change is based on @bvandermoon's [branch](https://github.com/AI-Hypercomputer/maxtext/compare/bvandermoon-nnx-single-kvcache-memory-issue) and keep it untouched for sanity check. This PR squashes all commits into one.

FIX: b/445953113, b/439070153

# Tests

Details in b/445953113#comment13

1. decode cmd: https://paste.googleplex.com/4995561203826688
- decode diff: https://diff.googleplex.com/#key=8kLLbM6jhJky, same as main (memstat after init, output)
2. train cmd: https://paste.googleplex.com/5801717128101888
- train diff: https://diff.googleplex.com/#key=9wKB5thwoIP9, same as main (TFLOP for last steps)
3. jetstream cmd, PDB=6: https://paste.googleplex.com/4992195518136320 (b/445953113#comment7)
- jetstream diff: https://diff.googleplex.com/#key=fqj8G4bz5Bom, same as main (acc, similar speed)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/MaxText/layers/nnx_wrappers.py [hunk 1]
```diff
@@ -163,12 +156,6 @@ def lazy_init(fn: Module | tp.Callable[..., tp.Any], *args, **kwargs):
   return fn
 
 
-def current_linen_module() -> linen.Module | None:
-  """Get the current Linen module from the Linen context."""
-  if linen.module._context.module_stack:  # pylint: disable=W0212
-    return linen.module._context.module_stack[-1]  # pylint: disable=W0212
-  return None
-
```
Comment (bvandermoon) [inline src/MaxText/layers/nnx_wrappers.py:171]: Could you double check if all the changes in this file are intentional? I had accidentally added extra changes on the branch I had. I think the changes in this file should just be all the ones for `to_linen_class`

Comment (RissyRan) [reply src/MaxText/layers/nnx_wrappers.py:171]: I didn't see this one is called. Let's discuss offline and see how to sanity check.

File: src/MaxText/maxengine.py [hunk 2]
```diff
@@ -114,7 +114,7 @@ def __init__(self, config: Any, devices: config_lib.Devices | None = None):
 
     # Model and Optimizer definition
     quant = quantizations.configure_quantization(config)
-    self.model = models.transformer_as_linen(config, mesh=self._mesh, quant=quant, model_mode=MODEL_MODE_PREFILL)
+    self.model = models.transformer_as_linen(config, mesh=self._mesh, quant=quant, model_mode=MODEL_MODE_AUTOREGRESSIVE)
```
Comment (bvandermoon) [inline src/MaxText/maxengine.py]: Why is this not still `MODEL_MODE_PREFILL`?

File: src/MaxText/maxtext_utils.py [hunk 3]
```diff
@@ -876,6 +876,7 @@ def init_initial_state(model, tx, config, is_training, key):
       {"params": key, "dropout": key, "aqt": key},
       np.ones(input_shape, dtype=jnp.int32),
       np.ones(input_shape, dtype=jnp.int32),
+      model_mode=MODEL_MODE_TRAIN,
```
Comment (bvandermoon) [inline src/MaxText/maxtext_utils.py]: Why is this `MODEL_MODE_TRAIN`? I see it is used in [setup_decode_state](https://github.com/AI-Hypercomputer/maxtext/blob/e9266c8e07d0aad0997afde6655b690056453905/src/MaxText/maxtext_utils.py#L887) if a [checkpoint is not loaded](https://github.com/AI-Hypercomputer/maxtext/blob/e9266c8e07d0aad0997afde6655b690056453905/src/MaxText/maxtext_utils.py#L1000-L1012). If you haven't already, can you also confirm this works for decode without loading a checkpoint?

Comment (bvandermoon) [reply src/MaxText/maxtext_utils.py]: Is this related to the way initializations been happening with model mode train in general?

File: src/MaxText/layers/nnx_wrappers.py (hunk 1 above)
Comment (RissyRan) [reply src/MaxText/layers/nnx_wrappers.py:171]: Verified by @shuningjin 

```
For nnx_wrapper.py, restore to main & copy our change in to_linen_class
unit test can pass
decode/train/jetstream looks fine: https://b.corp.google.com/issues/445953113#comment12
I commit this change to your PR directly, feel free to squash: https://screenshot.googleplex.com/6NDGK65yLvvuw4A
```

File: src/MaxText/maxengine.py (hunk 2 above)
Comment (RissyRan) [reply src/MaxText/maxengine.py]: I also remember this is `MODEL_MODE_PREFILL`. I recall I have verified it, but not sure why it gots autoregressive now.

File: src/MaxText/maxtext_utils.py (hunk 3 above)
Comment (RissyRan) [reply src/MaxText/maxtext_utils.py]: I recall this is needed; otherwise one unit tests will fail with missing `model_mode` or something.

Comment (shuningjin) [reply src/MaxText/maxtext_utils.py]: After removing [this line](https://github.com/AI-Hypercomputer/maxtext/pull/2370/commits/a43fa72a9e8608461e96bdddb0f5d3d1d9789d5a), the unit test is still green.

File: src/MaxText/maxengine.py (hunk 2 above)
Comment (shuningjin) [reply src/MaxText/maxengine.py]: I have [changed it back to MODEL_MODE_PREFILL](https://github.com/AI-Hypercomputer/maxtext/pull/2370/commits/883298ce6204178daf92d5698f07ba82c84e3097), the unit test is still green.

It seems to me the model_mode during initialization does not matter in our setting: https://screenshot.googleplex.com/BA5uNq6VDR8jPBR
because self.model_mode will be overridden by these "self.clone(model_mode=model_mode)" in init / apply:
https://screenshot.googleplex.com/6pfxLtdFJzZRJo8

**Review by shuningjin — APPROVED** [review]

**Comment by RissyRan** [comment]
Manually adding pull-ready label to trigger the G3 tests first.
